### ISSUE TITLE
Failed to run demo: Cannot read property 'ImageWrapper' of undefined

### ISSUE THREAD

**Issue by chenyong** [ericblade/quagga2#103]
I'm trying to run this demo in React, but always got error:

![image](https://user-images.githubusercontent.com/25790987/71579563-21981780-2b38-11ea-9fa1-b36334ea47dd.png)

```ts
import React, { useState, FC, useRef, useEffect, isValidElement } from "react";

import { QuaggaJSStatic } from "@ericblade/quagga2";
import QuaggaLib from "@ericblade/quagga2/lib/quagga";

let Quagga: QuaggaJSStatic = QuaggaLib;

let BarcodeArea: FC<{}> = (props) => {
  let videoElementRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    Quagga.init(
      {
        inputStream: {
          type: "LiveStream",
          target: videoElementRef.current,
          constraints: {
            width: 400,
            height: 400,
          },
        },
        locate: false,
        numOfWorkers: 2,
        frequency: 10,
        decoder: {
          readers: ["code_128_reader"],
        },
      },
      (err) => {
        if (err) {
          console.error(err);
          return;
        }

        console.log("started");
        Quagga.start();
      }
    );

    Quagga.onDetected((result) => {
      console.log("detect", result);
    });

    Quagga.onProcessed((result) => {
      console.log("process", result);
    });

    return () => {
      Quagga.stop();
    };
  }, []);

  return <div className={styleVideo} ref={videoElementRef}></div>;
};
```

Also I tried to run the demo at http://localhost:8080/example/live_w_locator.html by fixing the paths and adding `Quagga = quagga.default`. Turned out the problem remains.

Is there a fix for that?

**Comment by chenyong** [comment]
It looks like there is actually no `Quagga` defined in this function alone: https://github.com/ericblade/quagga2/blob/master/src/quagga.js#L377-L421

**Comment by Ridder90** [comment]
Facing the same issue here!

**Comment by ericblade** [comment]
Interesting question, I haven't looked at live_w_locator at all since i've started maintaining this fork .. 

I also haven't really looked at the worker interface much at all, at least in part due to the automated tests that I inherited not exercising it at all.  However, I do have a project that uses Quagga2 and uses the worker threads, and it hasn't stopped working (although i think it's still using 0.0.4 or something like that)

It looks like generateWorkerBlob() should be responsible for sending the worker some code to run, but it's not immediately obvious to me exactly what that does, and i'm not certain exactly when I'll get to being able to have a look at that, end of year craziness right now.

I'm not sure how React relates to live_w_locator -- live_w_locator doesn't use React, as far as I remember, and there's a separate example of React using Quagga2 here https://github.com/ericblade/quagga2-react-example/blob/master/package.json

that demo is currently using 0.0.7, but did work when i built it.  

Other thing that I can think of, is that there was a build problem with 0.0.11, might try updating to the 0.0.12, which i've tested all the way through the normal pipeline using the qr reader plugin tests.

I might get to having a look at live_w_locator in the next couple of days, might be a while, i'm not entirely certain.  If either of you come up with any further information, let me know :-)

oh, if you could provide exactly how you're running that demo, that'd be helpful too thanks :)

**Comment by chenyong** [comment]
> ...looked at live_w_locator...

I tried with the working demo on https://serratus.github.io/quaggaJS/examples/live_w_locator.html and looked into the code of workers. the worker code turned out to be much longer than the one I got. looks like it got a `Quagga` object there.

> I'm not sure how React relates to live_w_locator -- live_w_locator doesn't use React

my mistake, I tried the demo `live_w_locator`. also I tried with `@ericblade/quagga2` wrapped in my own React component(code paste above). Both of them run into the same problem `read property 'ImageWrapper' of undefined`. My current installed version is `@ericblade/quagga2@0.0.12`, guessing it should be the latest one.

Probably I can get my code bundled and put on web later today...

I'm feeling lucky Quagga still got a maintainer. Thanks for the work.

**Comment by chenyong** [comment]
BTW is there any other demo on your side which runs Quagga2 correctly?

**Comment by ericblade** [comment]
@chenyong the test suite in https://github.com/ericblade/quagga2-reader-qr builds and runs, i'm going to take a second and update the one here: https://github.com/ericblade/quagga2-react-example/blob/master/package.json

and see if that works.  i'll check back in a few minutes
If you're around, I'm on gitter, link in the README here.

**Comment by ericblade** [comment]
It's looking to me like it works.. at least in my browser, when served from localhost

![image](https://user-images.githubusercontent.com/1451847/71608974-c293cc00-2b52-11ea-8a57-2f197f18a5d5.png)

this is using @ericblade/quagga2^0.0.12

**Comment by chenyong** [comment]
I tried downloaded that react example as well, and upgraded to `0.0.12`,

![image](https://user-images.githubusercontent.com/25790987/71609459-93697300-2bc3-11ea-90e5-736462050748.png)

the camera(of macbook pro) worked. but when I open the Console, I see errors from worker threads again, on both Chrome Canary 80, and Chrome 79:

![image](https://user-images.githubusercontent.com/25790987/71609447-8056a300-2bc3-11ea-9d29-c8a264befda7.png)

**Comment by ericblade** [comment]
hmm. what's the value of navigator.hardwareConcurrency in Safari?

It does sound like the option to use workers is just straight up broken.  I know I was having issues with it trying to use it with custom readers, but I didn't even test it without a custom reader, i don't think..

I do see a bunch of the error on console on Chrome on Windows.  SO, I guess I've got a repro case.. 

I wonder if you might have time to go back through some previous revisions and see where it works correctly? if we can nail down a specific revision, we should be able to at least get a hint of what broke it.

My thinking is first try 0.0.2 then 0.0.3, as 0.0.3 was the version that introduced the external readers support, and is where i discovered external readers presently don't work with numOfWorkers set.  I hadn't even *considered* that i might've accidentally broken the worker feature completely, while i was trying to thread through the worker support for the external readers.

I'm going to be absolutely buried in work today / tomorrow (and then i have a lot of free time, that i'll be spending finding new work..) so if you have some time to try, lmk what you find out

**Comment by chenyong** [comment]
`undefined` in Safari.

![image](https://user-images.githubusercontent.com/25790987/71609902-1c35de00-2bc7-11ea-9c93-8281ca505715.png)

Also tried Chrome:

![image](https://user-images.githubusercontent.com/25790987/71609913-2ce65400-2bc7-11ea-824b-8cee1213cfab.png)

**Comment by chenyong** [comment]
Tried several versions, works with `0.0.10`, breaks in `0.0.11`.

**Comment by ericblade** [comment]
i wouldn't expect 0.0.11 to actually install, there's a build error there, npm tries to ... oh, i bet if you already had a different version there, it'd install the new one, before it failed out... 

I find that to be really interesting, that means that one of the typescript changes seems to be responsible, and I don't recall touching anything near the worker stuff.   ... i .. wonder if it's trying to ship straight up typescript code to the workers.  i need to learn how workers work.

use 0.0.10 in the meantime, i guess, or set numOfWorkers to 0

**Comment by chenyong** [comment]
> use 0.0.10 in the meantime, i guess, or set numOfWorkers to 0

yep. should be enough for our current need if it runs and return the code.

> wonder if it's trying to ship straight up typescript code to the workers. i need to learn how workers work. . . . and if it's shipping typescript code to the workers... how is it doing that?

not familiar with workers as well. which commit? I may have some time to see where it broke.

**Comment by ericblade** [comment]
I think I need to understand how it was that you examined the code in the workers, if you can point me to how to do that, it might help me understand.  i doubt that it's shipping typescript to the workers, but it might be shipping the contents of something that no longer exists.. that seems more likely.

anyway, my first idea would be to try doing a git bisect between 0.0.10 and 0.0.12, and that should point to the commit. there were a *lot* of commits in the typescript branch

**Comment by ericblade** [comment]
figured it out, it's the removal of the webpack-plugins\umd.js file that was removed as part of the upgrade to webpack from v2 to v5.  i.. don't think that can be directly put back in, but i think we'll need to study it.  it would put the entire source of the quagga.js module into the __factorySource__ variable .. and that no longer exists...

maybe it can just be put back in.. not sure...

**Comment by chenyong** [comment]
> it would put the entire source of the quagga.js module into the factorySource variable .. and that no longer exists...

I prettified code copied from Workers and noticed that code for QuaggaJS was missing. While there is quite some code(although minified so I was not sure) in the Workers of https://serratus.github.io/quaggaJS/examples/live_w_locator.html .

**Comment by ericblade** [comment]
ok, so.. i think we have a few possible options.. we could try to go back to doing it the way it was done before, or attempt to follow a guide like this: https://stackoverflow.com/questions/38715001/how-to-make-web-workers-with-typescript-and-webpack

or one of the links on that page points to this https://github.com/Qwaz/webworker-with-typescript

**Comment by chenyong** [comment]
Got some problem running `0.0.10`,

![image](https://user-images.githubusercontent.com/25790987/71610901-00830580-2bd0-11ea-84a8-f260664de458.png)

But it does work with `numOfWorkers: 0`.

I'll be out for a while. It's like already mid-night on your side.. Thanks for the help!

**Comment by ericblade** [comment]
fwiw, following the relatively simple instructions on the webpack worker-loader page at https://webpack.js.org/loaders/worker-loader/ I found it to be very easy to get a proper Worker setup.  Unfortunately, attempting to load Quagga into the Worker resulted in webpack infinite looping. 

SO, I think ultimately we need to restructure the worker handling code, which ultimately ties directly into restructuring the entire base of the library, to fix the well known problem where you can't decode multiple images simultaneously with decodeSingle().  

I think for the time being, I'm going to disable workers entirely

**Comment by chenyong** [comment]
sounds ok to us. not heavy use of barcode scanner at current.

**Comment by ericblade** [comment]
i'm not convinced that with the current structure of the worker, that there's even any benefit to it -- it just ran multiple copies of itself configured the same way, across multiple threads.  A restructuring of it might perhaps run different decoders on different worker threads.

I've published 0.0.13, if you want to try that and see if it fixes the issue.  I'm going to close this one anyway, in favor of the all-encompassing "restructuring things is really needed", but if this persists, lmk.. i didn't see it in my quick test (i did look at the console output this time :-D )
